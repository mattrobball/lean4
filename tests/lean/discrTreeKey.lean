universe u

def bar (_ _ : Nat) : Nat := default

#discr_tree_key (∀ {a n : Nat}, bar a (OfNat.ofNat n) = default)
#discr_tree_key (∀ {a n : Nat}, bar a (no_index OfNat.ofNat n) = default)
#discr_tree_key (∀ {a n : Nat}, bar a (no_index (OfNat.ofNat n)) = default)

variable (α : Type u) [Add α]

class AddSemigroup (α : Type u) extends Add α where
  add_assoc : ∀ (a b c : α), a + b + c = a + (b + c)

#discr_tree_key AddSemigroup.add_assoc

structure Wrapper (α : Type u) where
  val : α

namespace Foo

scoped instance foo : Add (Wrapper α) where
  add x y := ⟨x.val + y.val⟩

end Foo

namespace Bar

scoped instance bar : Add (Wrapper α) where
  add := fun ⟨x⟩ ⟨y⟩ => ⟨x + y⟩

end Bar

section

open Foo AddSemigroup

variable (β : Type u) [AddSemigroup β]

instance : AddSemigroup (Wrapper β) where
  add_assoc _ _ _ := congrArg Wrapper.mk (by
    discr_tree_key add_assoc
    discr_tree_key
    fail_if_success simp [add_assoc]
    sorry)

end

section

open Bar AddSemigroup

variable (β : Type u) [AddSemigroup β]

instance : AddSemigroup (Wrapper β) where
  add_assoc _ _ _ := congrArg Wrapper.mk (by
    have (b₁ b₂ b₃ : β) : b₁ + b₂ + b₃ = b₁ + (b₂ + b₃) := add_assoc _ _ _
    discr_tree_key this
    discr_tree_key
    simp [add_assoc])
end

