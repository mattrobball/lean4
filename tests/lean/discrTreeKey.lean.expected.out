bar _ (@OfNat.ofNat Nat _ _)
bar _ (@OfNat.ofNat Nat _ _)
bar _ _
@HAdd.hAdd _ _ _ _ (@HAdd.hAdd _ _ _ _ _ _) _
@HAdd.hAdd _ _ _ _ (@HAdd.hAdd _ _ _ _ _ _) _
@HAdd.hAdd β β β _ (@HAdd.hAdd (Wrapper β) (Wrapper β) (Wrapper β) _ x✝² x✝¹).1 x✝.1
discrTreeKey.lean:39:0-39:8: warning: declaration uses 'sorry'
@HAdd.hAdd β β β _ (@HAdd.hAdd β β β _ _ _) _
@HAdd.hAdd β β β _ (@HAdd.hAdd β β β _ x✝².1 x✝¹.1) x✝.1
